Remove SparseSeries and SparseDataFrame

TomAugspurger opened the pull request:
As discussed on the dev call.

Closes https://github.com/pandas-dev/pandas/issues/28407
Closes https://github.com/pandas-dev/pandas/issues/28025
Closes https://github.com/pandas-dev/pandas/issues/26777
Closes https://github.com/pandas-dev/pandas/issues/26227
~Not https://github.com/pandas-dev/pandas/issues/26123~
Closes https://github.com/pandas-dev/pandas/issues/25694
Closes https://github.com/pandas-dev/pandas/issues/25378
~Not https://github.com/pandas-dev/pandas/issues/25270~
~Not https://github.com/pandas-dev/pandas/issues/24817~
Closes https://github.com/pandas-dev/pandas/issues/24270
Closes https://github.com/pandas-dev/pandas/issues/22630
Closes https://github.com/pandas-dev/pandas/issues/21993
Closes https://github.com/pandas-dev/pandas/issues/21909
~Not https://github.com/pandas-dev/pandas/issues/21818~
Closes https://github.com/pandas-dev/pandas/issues/21231
Closes https://github.com/pandas-dev/pandas/issues/20766
Closes https://github.com/pandas-dev/pandas/issues/20172
~Not https://github.com/pandas-dev/pandas/issues/19278~
Closes https://github.com/pandas-dev/pandas/issues/18969
~Not https://github.com/pandas-dev/pandas/issues/18414~
Closes https://github.com/pandas-dev/pandas/issues/17993
Closes https://github.com/pandas-dev/pandas/issues/15737
Closes https://github.com/pandas-dev/pandas/issues/15634
Closes https://github.com/pandas-dev/pandas/issues/14427
Closes https://github.com/pandas-dev/pandas/issues/14310
Closes https://github.com/pandas-dev/pandas/issues/13665
~Not https://github.com/pandas-dev/pandas/issues/11726~
Closes https://github.com/pandas-dev/pandas/issues/10627
Closes https://github.com/pandas-dev/pandas/issues/9275
~Not https://github.com/pandas-dev/pandas/issues/9265~
Closes https://github.com/pandas-dev/pandas/issues/6427
Closes https://github.com/pandas-dev/pandas/issues/5470
Closes https://github.com/pandas-dev/pandas/issues/5078

TomAugspurger committed:
CLN: Remove sparse

TomAugspurger committed:
round 2

TomAugspurger committed:
Round 3

TomAugspurger committed:
round 4

TomAugspurger committed:
remove hdf

TomAugspurger committed:
some more

TomAugspurger committed:
cleanup

TomAugspurger committed:
note

jorisvandenbossche commented on pandas/io/pytables.py:
```diff
@@ -3078,83 +3069,6 @@ def write(self, obj, **kwargs):
-            kind=self.kind or "block",
-            fill_value=self.fill_value,
-            name=self.name,
-        )
```
I don't know exactly how pytables IO works, but in principle we could adjust this to read existing data as a Series[sparse]

TomAugspurger replied:
Yeah I was wondering about that. We do have an issue about supporting EAs in IO (HDF5 in particular). How should we handle this?

I suppose the best thing to do is to change those to return Series[sparse] before this PR?

jorisvandenbossche replied:
That's also an option. But I don't think we need to try to solve the full "EA in HDF5" problem before this PR, since sparse is actually already supported.

jorisvandenbossche replied:
Can take a look at that

TomAugspurger replied:
> since sparse is actually already supported.

You mean `Series[sparse]` works already? I didn't think it did.

jorisvandenbossche replied:
No, I meant that SparseSeries already did, but since Series[sparse] is almost an identical replacement (regarding what data is stored in the Series), it should be rather easy to get Series[sparse] working (while getting EAs in general working might be more effort)

TomAugspurger replied:
To close this subitem, reading SparseSeries / SparseDataFrame files is broken on master. So HDF5 support isn't a blocker.

jorisvandenbossche commented on pandas/tests/io/generate_legacy_storage_files.py:
```diff
@@ -53,10 +53,7 @@
     Period,
     RangeIndex,
     Series,
-    SparseDataFrame,
```
I think you can leave this file intact for now? 
(isn't it meant to be run with older version of pandas? In case we need to regenerate old pickle files, we might still want to generate sparse ones to check that reading them works / errors with a good message)

TomAugspurger replied:
IIUC, we need to remove it so that the sparse structures are not included, re-run it with an older pandas, and commit the new binary file.

I will update to use SparseArray though.

TomAugspurger replied:
@WillAyd can you comment here? I think we import this file when running the tests, so we need to be able to import it under both the old and new pandas?

jorisvandenbossche replied:
Should this file now be restored now you added the pickle compat? (I also think you now no longer update the old pickle files, right?)

TomAugspurger replied:
I think this file gets imported on master, so I need to remove this.

jorisvandenbossche replied:
if so, you can put it behind a try/except?

TomAugspurger replied:
That alone probably isn't what we want though, right? Since even if it's in a try / except, it's not actually being used.

I think the best thing is to include newly generated legacy files, in addition to the old ones. Then we verify that the new legacy files (with Series[sparse] rather than SparseSeries, etc.) are working, and that we're still able to read old files generated from SparseSeries / SparseDataFrame.

jorisvandenbossche replied:
I am not very familiar with how this is used in practice, but I thought it is something like:

- we decide to raise minimum supported versions on read from 0.20 to 0.23 (we just bumped to minimum of 0.20 in 0.25)
- we run this script with pandas 0.23 to generated the new "legacy pickle files" and replace the old ones with those

If that is correct (@WillAyd is that what you did for 0.25?), then the script still needs to be able to generate pickles files with SparseDataFrame (in this hypothetical case, if we would directly bump to eg 1.0, we wouldn't need it)

WillAyd replied:
Sorry I missed the original ping. I haven't been able to check in detail but will carve out some time later today. If it helps in the interim here is where I cleaned up the old compat files for Panel:

https://github.com/pandas-dev/pandas/pull/27082

WillAyd replied:
The changes here are correct; you just need to re-run this in an environment with pandas 0.20.3 ideally using the minimum supported Python version, which I guess right now is 3.5.3 though by the time this goes live should be 3.6, so maybe stick to the latter (I see there is one for 3.5.6 already)

jorisvandenbossche replied:
@WillAyd can you answer on my comment above in this thread? (https://github.com/pandas-dev/pandas/pull/28425/files#r324859383) Why would you not want to recreate the pickle files *with* SparseDataFrame? Those legacy pickle files are meant to represent the legacy situation, no? And in 0.20.3, those classes exist, and can be written to pickle files.

WillAyd replied:
I don’t know the entire history of these files either so may be missing intent, but if we don’t have a SparseDataFrame anymore then there wouldn’t be a reason to try and read one as that object can’t be recreated on master, no?

jorisvandenbossche replied:
But this file is not mean to recreate them with master, but with an older release? (or is that the part I am misunderstanding? You ran that script with pandas 0.21, not with master?)

WillAyd replied:
Right the file would be created with an old release but read and tested against master I think

jorisvandenbossche replied:
So if we want to mimic what users can do and what Tom is trying to support (i.e. write a pickle file with SparseDataFrame in an older release and read this file with the new release), this file needs to be able to generate such pickle files with SparseDataFrame?

Now, I don't really care that much about SparseDataFrame, and there are currently already legacy pickle files that contain it, so the need to be able to recreate those files before we will actually remove that support is maybe not that high, and I am maybe splitting hairs.

TomAugspurger replied:
Sorry @jorisvandenbossche, I think I've misunderstood you this whole time (and I was confused why you kept pushing on this point 😄). You're correct that if we relied on these legacy files for testing reading files written with SparseSeries & SparseDataFrame, then we'd need to keep the imports here (at conditionally).

But https://github.com/pandas-dev/pandas/pull/28425/commits/a8b0d65154b76fcb0b62025d87d8ead9fcdde723 added tests dedicated for that (we need dedicated tests, since the existing ones silence all warnings).

So from a test coverage POV, we're OK with removing writing SparseSeries & SparseDataFrame here. It's just a matter of code style / prep for running this file when 1.0.0 becomes the "legacy" pickle version we support.

jorisvandenbossche replied:
> So from a test coverage POV, we're OK with removing writing SparseSeries & SparseDataFrame here.

Yep, if we have files now, it seems ok to remove it from this file. It seems rather unlikely that after the next bump in legacy support for pickle files, we still want to keep support for SparseDataFrame, I suppose, as we now warn for it?

TomAugspurger replied:
That sounds right to me. Technically, we could chose to bump the next legacy to version 0.25.x, which has SparseDataFrame, but I think we're more likely to choose 1.0.x since it's a round number.

jorisvandenbossche commented on pandas/tests/io/json/test_pandas.py:
```diff
@@ -1296,25 +1296,6 @@ def test_datetime_tz(self):
-        df = pd.DataFrame(np.random.randn(10, 4))
-        df.loc[:8] = np.nan
-
-        sdf = df.to_sparse()
```
I think you can actually do `sdf = df.astype('Sparse')` and for the rest keep this test? (so we test sparse series that way as well, not sure it is covered somewhere else)

jorisvandenbossche commented on pandas/tests/series/test_combine_concat.py:
```diff
@@ -273,39 +271,6 @@ def test_concat_empty_series_dtypes(self):
-        # sparse
-        # TODO: move?
-        result = pd.concat(
-            [Series(dtype="float64").to_sparse(), Series(dtype="float64").to_sparse()]
```
Here can also do the astype thing instead of to_sparse

jorisvandenbossche reviewed (commented):
You can do a search for `@pytest.mark.filterwarnings("ignore:Sparse:FutureWarning")`, I thnk there are a few left (or at least at the moment I was reviewing)

TomAugspurger commented on pandas/core/ops/__init__.py:
```diff
@@ -32,7 +32,6 @@
     ABCExtensionArray,
     ABCIndexClass,
     ABCSeries,
-    ABCSparseSeries,
```
Yes, I need to run coverage and see what's dead now.

jorisvandenbossche replied:
My text search in the full code base said they were not used anywhere

TomAugspurger commented:
@WillAyd it looks like you fixed up the pickle panel stuff in https://github.com/pandas-dev/pandas/pull/27082/files.

Would you recommend removing SparseSeries / SparseDataFrame from the pickle files as a precursor to this PR?

TomAugspurger commented on pandas/core/sparse/scipy_sparse.py:
```diff
@@ -1,7 +1,7 @@
 """
 Interaction with scipy.sparse matrices.
 
-Currently only includes SparseSeries.to_coo helpers.
```
Yeah, I was going to do moves as a followup. I'd also like to move the last bit of `pandas/tests/sparse/test_combine_compat.py`, which had a SparseArray test in it.

TomAugspurger commented on pandas/core/sparse/api.py:
```diff
@@ -1,4 +1,3 @@
-# flake8: noqa
 from pandas.core.arrays.sparse import SparseArray, SparseDtype
-from pandas.core.sparse.frame import SparseDataFrame
-from pandas.core.sparse.series import SparseSeries
```
The file? Dunno. We do still import from it in some tests (should be using the public API).

I suppose this falls under the `.core` is private discussion. I'm OK with leaving it for now.

TomAugspurger committed:
fixups

TomAugspurger committed:
pickle changes

TomAugspurger commented:
3cc47659dc16f57370bc8e4a82463ec69fae7dc0 has the changes to the pickle stuff. I can split that out into a precursor PR if people want.

I regenerated the pickle files with python 3.5.6 and pandas 0.20.3.

jorisvandenbossche commented:
Do we still have pickle files with sparse data in them? 
What error do you get if you read such a file with this branch?

TomAugspurger commented:
> Do we still have pickle files with sparse data in them?
> What error do you get if you read such a file with this branch?

As of https://github.com/pandas-dev/pandas/commit/3cc47659dc16f57370bc8e4a82463ec69fae7dc0 we have a pickle file with Series[Sparse] and DataFrame[Sparse], but no SparseSeries and SparseDataFrame. That converted the use of SparseSeries/Frame to Series/Frame with sparse values.

With https://github.com/pandas-dev/pandas/commit/3cc47659dc16f57370bc8e4a82463ec69fae7dc0, the pickle tests on this branch pass. It would also pass on master, so I could split those changes out.

jorisvandenbossche commented:
I don't have a strong opinion on splitting or not, either way.

For older pickle files with SparseDataFrame in it, you currently get:

```
AttributeError: Can't get attribute 'SparseDataFrame' on <module 'pandas.core.sparse.frame' from '/home/joris/scipy/pandas/pandas/core/sparse/frame.py'>
```

With our pickle_compat, I suppose we could actually map this to a function that creates a DataFrame[sparse] of it. But not sure that's worth the effort. 
But a better error message indicating they need to read it in 0.25, convert and save again to read it in 1.0 would be welcome I think.

TomAugspurger committed:
pickle compat

TomAugspurger committed:
skip feather

TomAugspurger committed:
fixups

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into remove-sparse

TomAugspurger commented:
Huh, 766a2f2b1cd6ee53b0a8124ca04aac3460ad66fd has a bit of compat code to allow reading a file saved with SparseSeries / SparseFrame. I don't really know how it works, but it seems to. I've restored the old pickle file with SparseSeries / SparseFrame and it passes locally.

TomAugspurger committed:
cleanups

TomAugspurger committed:
black

TomAugspurger committed:
to_sparse docs

TomAugspurger committed:
doc note

TomAugspurger committed:
rm sparse frame

TomAugspurger committed:
rm sparse series

TomAugspurger committed:
docs

TomAugspurger committed:
doc

TomAugspurger committed:
remove new pickle

TomAugspurger commented:
This will hopefully pass. To summarize a few things

- I'm skipping the majority of the Feather reader / writer tests because of users would have to add another dependency (e.g conda install pandas-stata) and possibly an extra import in their code
- We're still able to read old pickle files with SparseSeries or SparseDataFrame because of the loaders I added in pickle_compat
- We're semi-blocked by https://github.com/pandas-dev/pandas/pull/28425#discussion_r324212982 (HDF5 io). IMO, we can merge this first and restore sparse IO for HDF5 later (but before 1.0), but either works for me.

TomAugspurger committed:
Update v0.25.0.rst

TomAugspurger committed:
Update v0.25.0.rst

TomAugspurger commented:
@jorisvandenbossche did you say you wanted / could look into HDF5 serialization? If not I can take care of it.

I looked at things briefly, and it seems pretty doable. We should be able to read existing HDF5 files w/ sparse data, and write news using most of the same code.

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into remove-sparse

jorisvandenbossche commented:
I am working on the doc theme / geopandas stuff today, so if you want to do it today, go ahead!

jorisvandenbossche reviewed (commented):
This looks good to me.

For the docs, we need to see if it is needed to add actual output to the code-blocks

For the issues you are closing with this, we actually need to see if some of them are not also valid for Series[sparse] / SparseArray

jorisvandenbossche commented:
> For the issues you are closing with this, we actually need to see if some of them are not also valid for Series[sparse] / SparseArray

I already went through the most recent ones, and crossed a few in the top post.

TomAugspurger committed:
added new legacy pickle files

WillAyd commented on pandas/tests/io/generate_legacy_storage_files.py:
```diff
@@ -86,6 +84,12 @@
     YearEnd,
 )
 
+try:
```
Do you know which version this was changed in? Would be preferable to change import based on pandas version so we know when to clean up

WillAyd commented on doc/source/user_guide/io.rst:
```diff
@@ -4658,26 +4658,41 @@ See the `Full Documentation <https://github.com/wesm/feather>`__.
 
-   result = pd.read_feather('example.feather')
-   result
+.. code-block:: python
```
Just out of curiosity any reason this isn't an ipython block? Seems like it could reduce code required

jorisvandenbossche replied:
Because released feather (pyarrow) fails with this PR due to the removed SparseDataFrame (in pyarrow master and upcoming 0.15 release this is already fixed). 

That reminds: Tom, we discussed adding a dummy SparseDataFrame as we did for Panel to avoid direct breakage like this?

TomAugspurger replied:
I don't recall that's a good idea. Then we're not losing a bunch of our feather test coverage while waiting for the next pyarrow release.

WillAyd commented on doc/source/whatsnew/v0.16.0.rst:
```diff
@@ -91,8 +91,7 @@ Interaction with scipy.sparse
 
-.. ipython:: python
-   :okwarning:
+.. code-block:: python
```
Could be done as a follow up and probably something easy for the community to contribute to, but ideally since we are converting these from `ipython` to `python` directives should add the original output (applicable in a few places below)

TomAugspurger replied:
Certainly wouldn't object, but I think that's fairly low-value since, e.g. the 0.16 version of the docs at https://pandas.pydata.org/pandas-docs/version/0.16.0/whatsnew.html will have the correct rendering.

WillAyd replied:
Yea I don’t think you should do here as there’s already enough - just a nice to have a first time contributor could pick up. 

I can open an issue for it later

WillAyd commented on pandas/compat/pickle_compat.py:
```diff
@@ -54,6 +55,22 @@ def load_reduce(self):
         raise
 
 
+class _LoadSparseSeries:
```
I'm not super familiar with this module but is this definitely something we want to do? Implicitly convert `SparseSeries` from old files to just a `Series`?

jorisvandenbossche replied:
I think this is something we want to do, to provide some kind of backwards compatibility on older pickle files (note that the output will still be sparse, just not the subclass), and given how easy it is to provide this. 
But, I think we should add a warning to say that the SparseDataFrame has been converted to DataFrame with sparse columns, and to warn this will not be supported anymore in the future (so they need to re-save the files with the new version).

TomAugspurger replied:
Will do.

WillAyd reviewed (commented):
Generally nice! This is quite the clean up. Just some comments to consider but I think this looks really good

WillAyd commented on pandas/compat/pickle_compat.py:
```diff
@@ -54,6 +55,22 @@ def load_reduce(self):
 
+class _LoadSparseSeries:
+    # To load a SparseSeries as a Series[Sparse]
+    def __new__(cls) -> Any:
```
Also w.r.t annotation could probably just use `FrameOrSeries` from pandas._typing, as that is tied to subclasses of NDFrame (applicable to both methods)

jorisvandenbossche commented:
> added new legacy pickle files 

Maybe rather add files for a different (later) version, since we already have them for 0.21?

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into remove-sparse

TomAugspurger committed:
shim

TomAugspurger committed:
shim

TomAugspurger committed:
revert io changes

TomAugspurger committed:
warning for sparse

TomAugspurger committed:
Fixup typing

TomAugspurger committed:
format

TomAugspurger committed:
fixup typing

TomAugspurger committed:
0.24.0 todo

TomAugspurger commented:
Just FYI, since the last round of reviews I've added

1. A dummy `SparseDataFrame` and `SparseSeries` class, mirroring what we did for Panel. This prevents the immediate error with feather (I've removed the test xfails and reverted the doc changes working around it)
2. Added a warning when loading a pickle file with a SparseDataFrame / SparseSeries saying they should re-save it

jorisvandenbossche reviewed (approved):
Last changes look good to me!

TomAugspurger commented:
Going through the rest of the issues to validate that they are specific to the sparse subclasses.

jorisvandenbossche commented:
Let's merge this?

jreback commented:
lgtm. I am -0 on providing the pickle compat. but I guess ok.

jreback commented:
merge when ready.

TomAugspurger commented:
Gave another glance through, and I think we're good.

Mostly agreed about the pickle compat. But since it turned out to be so easy (same internal representation) I think we're OK.

jorisvandenbossche commented:
Woohoo! :)

>  I am -0 on providing the pickle compat. but I guess ok.

I think a reason to have it (apart from that it was easy to add) is because this is a very recent deprecation, and that it also maps nicely to a supported data structure.
